storage/aptosdb/src/pruner/state_store/mod.rs

```diff
@@ -102,140 +85,119 @@ impl DBPruner for StateStorePruner {
+                batch.delete::<JellyfishMerkleNodeSchema>(&index.node_key)?;
+                batch.delete::<StaleNodeIndexSchema>(&index)
+            })?;
+            // Delete the stale node indices.
+            self.db.write_schemas(batch)?;
+            self.pruning_completed_version
+                .store(pruning_completed_version, Ordering::Relaxed);
+            Ok(new_min_readable_version)
+        }
     }
 
-    fn next_result(&mut self) -> Result<Option<Vec<StaleNodeIndex>>> {
-        match self.inner.next().transpose()? {
-            None => Ok(None),
-            Some((index, _)) => {
-                let version = index.stale_since_version;
-                if version > self.target_min_readable_version {
-                    return Ok(None);
-                }
+    // Return the stale node indices to prune in one iteration. It either returns `batch_size` of
+    // items or all the remaining items whose stale_since_version = target version. It also returns
+    // the last version which all the stale nodes has been pruned.
+    fn get_stale_node_indices(
+        &self,
+        start_version: Version,
+        target_version: Version,
+        batch_size: usize,
+    ) -> Result<(Vec<StaleNodeIndex>, Version)> {
+        let mut indices = Vec::new();
+        let mut iter = self
+            .db
+            .iter::<StaleNodeIndexSchema>(ReadOptions::default())?;
+        iter.seek(&start_version)?;
 
-                let mut indices = vec![index];
-                while let Some(res) = self.inner.peek() {
-                    if let Ok((index_ref, _)) = res {
-                        if index_ref.stale_since_version != version {
+        let mut num_items = batch_size;
+        let mut pruning_completed_version = if start_version == 0 {
+            0
+        } else {
+            start_version - 1
+        };
+        while num_items > 0 {
+            if let Some(item) = iter.next() {
+                match item {
+                    Err(e) => {
+                        return Err(e);
+                    }
+                    Ok((index, _)) => {
+                        if index.stale_since_version > target_version {
+                            pruning_completed_version = target_version;
                             break;
                         }
+                        num_items -= 1;
+                        indices.push(index);
                     }
-
-                    let (index, _) = self.inner.next().transpose()?.expect("Should be Some.");
-                    indices.push(index);
                 }
-
-                Ok(Some(indices))
+            } else {
+                // No more stale nodes.
+                break;
             }
         }
-    }
-}
-
-impl<'a> Iterator for StaleNodeIndicesByVersionIterator<'a> {
-    type Item = Result<Vec<StaleNodeIndex>>;
 
-    fn next(&mut self) -> Option<Self::Item> {
-        self.next_result().transpose()
+        if let Some(next_item) = iter.next() {
```

Comment:
comment here this only deal with the case where batch size is met, but iterator can still have next item.